Round source scale to nearest integer if very close.

petewa opened the pull request:
### Reason for this pull request

Follow up PR to #1047 to snap scale to nearest integer


### Proposed changes

- Snap scale to nearest integer in `box_overlap`
- Bring `maybe_int` and `maybe_zero` from `odc-tools`



 - [ ] Closes #xxxx
 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

<!--
See https://github.com/blog/2111-issue-and-pull-request-templates for more
information on pull request templates.

-->

Kirill888 commented on datacube/utils/geometry/tools.py:
```diff
@@ -446,20 +446,24 @@ def box_overlap(src_shape, dst_shape, ST, tol):
-        if is_almost_int(x, tol):
-            x = round(x)
-        return x
+    def snap_scale(s, tol=1e-6):
```
Can we move this out of this function and add tests for it?
Remember that scale can be negative,  this case should be handled properly.

Kirill888 commented on datacube/utils/geometry/tools.py:
```diff
@@ -446,20 +446,24 @@ def box_overlap(src_shape, dst_shape, ST, tol):
+        if s > 1:
+            s = maybe_int(s, tol)
+        elif s < 1:
+            s = 1.0/maybe_int(1.0/s, tol)
```
This needs to be done differently, it should return exactly the same value as passed in when not snapping to a fraction.

`maybe_int` returns exactly the same object when snapping is not happening, so if `maybe_int(inv_s) is inv_s` then you just return `s`, and only if `inv_s` was modified do you change `s`.

petewa committed:
Round source scale to nearest integer if very close.
This fixes compute_axis_overlap asking for 1 more pixel than it should
in certain situations where scale is very near an integer but not exactely.
e.g. 0.9999999999999747 or 1.0000000000000253

Kirill888 commented on datacube/utils/math.py:
```diff
@@ -69,6 +69,46 @@ def spatial_dims(xx: Union[xr.DataArray, xr.Dataset],
+    """ Snap scale to to nearest integer.
+    """
+    maybe_int_s = maybe_int(s, tol)
+    if maybe_int_s == s:
+        return s
+    else:
+        return maybe_int_s
```
This is redundant, `maybe_int` already ensures that if there is no rounding then input is returned exactly as it was passed in, my previous comment was about handling snapping to `1/<some_int>`

petewa replied:
I'll revert it back to how I originally had it before all these changes with just using maybe_int_s. I don't think we need to have snap_scaled at all. Since we are targeting only snapping to integers.

But snapping it to a simple fraction is good too.

Kirill888 commented on tests/test_utils_other.py:
```diff
@@ -541,6 +544,31 @@ def test_is_almost_int():
+
+
+def test_maybe_int():
+    assert maybe_int(1, 1e-10) == 1
```
Needs tests that verify that "same object" is returned when there is no truncation

```python
for x in [3/7, 7/3, -13.7878]:
    assert maybe_int(x) is x
```

petewa replied:
I'll add this in

petewa committed:
Modified snap_scale to handle
  - snapping to near integer
  - snapping to simple fractions
  - return scale unchanged is not snapping.
Added test for maybe_int to verify that "same object" is returned when
there is no truncation

petewa committed:
Added an extra check, tests and removed some redundant code.

petewa committed:
Modified snap_scale logic so it's correct.

Kirill888 commented on datacube/utils/math.py:
```diff
@@ -69,6 +69,54 @@ def spatial_dims(xx: Union[xr.DataArray, xr.Dataset],
+
+
+def snap_scale(s, tol=1e-6):
+    """ Snap scale to the nearest integer and simple fractions in the from 1/s
```
```suggestion
    """ Snap scale to the nearest integer and simple fractions in the form 1/<int>
```

Kirill888 commented on datacube/utils/math.py:
```diff
@@ -69,6 +69,54 @@ def spatial_dims(xx: Union[xr.DataArray, xr.Dataset],
+        return maybe_int(s, tol)
+    else:
+        # Check of s is 0
+        if s == 0:
```
```suggestion
        if abs(s) < tol:
```

Kirill888 committed:
More robust checking for 0 + spelling corrections
